dsl/src/commonMain/kotlin/it/unibo/collektive/field/Field.kt

```diff
@@ -104,27 +104,78 @@ sealed interface Field<ID : Any, out T> {
+            hoodWithId(default) { (_, accumulator), (id, value) -> id to transform(accumulator, value) }
+
+        /**
+         * Reduce the elements of the field using the [transform] function,
+         * if the (neighboring) field is empty the [default] value is returned.
+         * The [transform] function takes two pairs: the former represents the accumulated value (including the [ID]),
+         * while the latter represents the current entry of the neighboring field that should be combined.
+         *
+         * Use this function when the [ID] should be propagated during the reduce operation.
+         */
+        inline fun <ID : Any, T> Field<ID, T>.hoodWithId(
+            default: T,
+            crossinline transform: (Pair<ID, T>, Pair<ID, T>) -> Pair<ID, T>,
+        ): T {
             val neighbors = excludeSelf()
             return when {
                 neighbors.isEmpty() -> default
-                else -> neighbors.values.reduce(transform)
+                else ->
+                    neighbors.entries
+                        .asSequence()
+                        .map { it.toPair() }
+                        .reduce { accumulator, value -> transform(accumulator, value) }
+                        .second
             }
         }
 
         /**
-         * Folds the elements of a field starting with an [initial] through a [transform] function.
-         * The local value is not considered, unless explicitly passed as [initial].
+         * Reduce the elements of the field using the [transform] function,
+         * it includes the [ID] of the element whenever it should be considered in the [transform] function,
+         * but the [ID] is not returned.
+         * The local value of the field is not considered.
+         * Returns the [default] if the field to transform is empty.
+         */
+        inline fun <ID : Any, T> Field<ID, T>.hoodWithId(default: T, crossinline transform: (T, ID, T) -> T): T =
+            hoodWithId(default) { (_, accumulator), (id, value) -> id to transform(accumulator, id, value) }
+
+        /**
+         * Accumulates the elements of a field starting from an [initial] through a [transform] function.
+         * The local value of the field is not considered.
          */
-        fun <ID : Any, T, R> Field<ID, T>.fold(initial: R, transform: (R, T) -> R): R {
+        inline fun <ID : Any, T, R> Field<ID, T>.fold(initial: R, crossinline transform: (R, T) -> R): R =
+            foldWithId(initial) { accumulator, (_, value) -> transform(accumulator, value) }
+
+        /**
+         * Accumulates the elements of a field starting from an [initial] through a
+         * [transform] function that includes the [ID] of the element.
+         * The local value of the field is not considered.
+         */
+        inline fun <ID : Any, T, R> Field<ID, T>.foldWithId(
+            initial: R,
+            crossinline transform: (R, Pair<ID, T>) -> R,
```

Comment:
For consistency with `hoodWithId` why `transform` is not of the kind `(Pair<ID, R>, Pair<ID, T>) -> Pair<ID, R>`?